firewall/inspection/portscan/detect.go

```diff
@@ -0,0 +1,357 @@
+package portscan
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/safing/portbase/log"
+	"github.com/safing/portbase/modules"
+	"github.com/safing/portmaster/firewall/inspection"
+	"github.com/safing/portmaster/netenv"
+	"github.com/safing/portmaster/network"
+	"github.com/safing/portmaster/network/netutils"
+	"github.com/safing/portmaster/network/packet"
+	"github.com/safing/portmaster/process"
+	"github.com/safing/portmaster/status"
+)
+
+type tcpUDPport struct {
+	protocol packet.IPProtocol
+	port     uint16
+}
+
+type ipData struct {
+	score int //score needs to be big enough to keep maxScore + addScore... to prevent overflow
+	//	greylistingWorked bool
+	previousOffender bool
+	blocked          bool
+	ignore           bool
+	lastSeen         time.Time
+	lastUpdated      time.Time
+	blockedPorts     []tcpUDPport
+}
+
+const (
+	//fixme: Which production-values do we want to have?
+	cleanUpInterval = 1 * time.Minute //fixme: Debug-Value
+	cleanUpMaxDelay = 5 * time.Minute
+
+	startAfter            = 1 * time.Second //fixme: Debug Value; When should the Portscan Detection start to prevent blocking Apps that just try to reconnect?
```

Comment:
The scenario is: Outbound-connection before reboot -> reboot; data still coming in at the old closed socket.
Maybe we can make exclude only dynamic ports at startup :thinking: 

How do we check if it incoming or outgoing? We lookup if an app is connected to the port. At reboot, all the app connections are gone ...

Reply:
How does it check if the connection is incoming? If there is an app listening on that port.

The scenario is: Data beeing sent on a socket, the PC reboots. The other device is still sending data on the old socket(s).

Reply:
:D OK I didn't think about that. Yeah, not sure this is really a problem. Did you observe the problem in the wild or is it more a theoretical issue? IMHO this could only be an issue for UDP because a TCP connection would have died already or the kernel would immediately send a RST so there would be at most one detected "port probe". For UDP that may be correct but in order to be rated as a portscan that other "device" must send to lots of different ports. I doubt that's really going to happen.
Also, there was enought time for the whole PC to reboot so IMHO one second more won't change a lot here.

Reply:
theoretical Problem

Reply:
Hope that is no problem in practice  :thinking: